Geochem: lengthy and fundamental additions

WilkAndy opened the pull request:
This stuff will be used substantially in the remainder of the geochem code, and may have to be refactored as we progress.  At the moment it is just unit-tested utilities, but i have MOOSE wrappers that give a nice user interface - i'll push that up with documentation, input files, etc, next week. 

Gird your loins, @cpgr and @loganharbour - i've been working on this for a while and it is nontrivial.

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

WilkAndy committed:
fixed unit of surface potential in documentation.  Fixes #15243

WilkAndy committed:
GeochemistrySpeciesSwapper does not allow equilibrium species that are involved in surface sorption to be swapped into the basis.  Fixes #15245

WilkAndy committed:
Constants added to geochemistry that are relevant to surface sorption.  Fixes  #15246

WilkAndy committed:
MinimalGeochemicalDatabase now holds information related to surface sorption.  Fixes #15247

WilkAndy committed:
extra sorption info added to unit/database/moose_testdb.json .   Refs #15247 .    Fixes ##15244

WilkAndy committed:
when equlibrium radius = -1.5, activity coefficient = 1.0.  Refs #15247

WilkAndy committed:
EquilibriumGeochemicalSystem added.  At present this satisfies my requirements, but may require refactoring later since this IS A KEY CLASS IN THE GEOCHEMISTRY MODULE.  Fixes # #15260

WilkAndy committed:
Solver added to solve 0D equilibrium problems.  Fixes  #15261

WilkAndy commented:
Ooops, something wrong with MOOSE (presumably not caused by this PR):
```
/opt/civet/build_0/moose/framework/build/header_symlinks/MooseEnum.h:31:7: warning: unused parameter 'other_enum' [-Wunused-parameter]
In file included from /opt/civet/build_0/moose/framework/build/header_symlinks/DistributionInterface.h:12:0,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/UserObject.h:13,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/SideUserObject.h:13,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/SidePostprocessor.h:12,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/SideIntegralPostprocessor.h:12,
                 from /opt/civet/build_0/moose/modules/heat_conduction/build/header_symlinks/ConvectiveHeatTransferSideIntegral.h:3,
                 from /opt/civet/build_0/moose/modules/heat_conduction/src/postprocessors/ConvectiveHeatTransferSideIntegral.C:1,
                 from /opt/civet/build_0/moose/modules/heat_conduction/build/unity_src/postprocessors_Unity.C:2:
/opt/civet/build_0/moose/framework/build/header_symlinks/InputParameters.h:1244:11: note: synthesized method 'MooseEnum& MooseEnum::operator=(const MooseEnum&)' first required here 
   l_value = r_value;
           ^
```

WilkAndy commented:
Hi @rpodgorney an @cpgr ,

I'm quite happy with the progress so far, although there is a lot to go!  This PR allows users to fully solve zero-dimensional equilibrium reaction systems.  Might not sound a lot, but when you look at the lines of code you'll see it is.

You can solve equilibrium systems including those with redox decoupling and surface complexation.  Examples are:
https://mooseframework.inl.gov/modules/geochemistry/tests_and_examples/seawater.html
https://mooseframework.inl.gov/modules/geochemistry/tests_and_examples/amazon.html
https://mooseframework.inl.gov/modules/geochemistry/tests_and_examples/red_sea.html
https://mooseframework.inl.gov/modules/geochemistry/tests_and_examples/morro.html
https://mooseframework.inl.gov/modules/geochemistry/tests_and_examples/microbial_redox.html
https://mooseframework.inl.gov/modules/geochemistry/tests_and_examples/gypsum.html
https://mooseframework.inl.gov/modules/geochemistry/tests_and_examples/sebkhat.html
https://mooseframework.inl.gov/modules/geochemistry/tests_and_examples/selenate.html
https://mooseframework.inl.gov/modules/geochemistry/tests_and_examples/surface_complexation.html

in some of those examples, i get different results from Bethke.  I'm not sure whether that due to the lack of temperature least-squares fit in `geochemistry` currently (Chris is addressing this) or something else (usually differences are < 1%, but in the sorption stuff they are around 50% for some quantities!).  But the good news is that i'm going a GeochemistsWorkbench course later this month, that includes a 60-day copy of GWB, so i can check out all discrepancies very thoroughly.

This PR is quite big and i'll be adding to it next week.  It may take a few days to review and merge.  After this i'll be on to:

- time dependent things (changing temperature, progressively adding chemicals, changing pH, removing minerals, etc)
- kinetics
- transport!

a

cpgr commented:
You are amazing Andy! I'll try and take a look at it this afternoon.

rpodgorney commented:
Wow Andy!!! Amazing. I’ll have a look next week.

Robert Podgorney, PhD, PG
Lead, Geothermal and Fossil Energy
Energy and Environment Division
Idaho National Laboratory

Office: 208-526-1524<tel:208-526-1524>
Cell: 208-520-9361<tel:208-520-9361>
email: robert.podgorney@inl.gov<mailto:robert.podgorney@inl.gov>

-sent from my mobile device, please excuse any typos or autocorrect errors

Statement of Confidentiality
The contents of this e-mail message and any attachments are confidential and are intended solely for addressee. The information may also be legally privileged. This transmission is sent in trust, for the sole purpose of delivery to the intended recipient. If you have received this transmission in error, any use, reproduction or dissemination of this transmission is strictly prohibited. If you are not the intended recipient, please immediately notify the sender by reply e-mail or phone and delete this message and its attachments, if any.


On May 7, 2020, at 20:44, Chris Green <notifications@github.com> wrote:

﻿

You are amazing Andy! I'll try and take a look at it this afternoon.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://gcc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fidaholab%2Fmoose%2Fpull%2F15262%23issuecomment-625598041&data=02%7C01%7Crobert.podgorney%40inl.gov%7C85a17b2bfe9b49eb0e5a08d7f2f9bac2%7C4cf464b7869a42368da2a98566485554%7C0%7C0%7C637245026717829249&sdata=IX2FCt6xhlLPUEvlofK5TwDG3n3JPtPcRlNbij4gPJw%3D&reserved=0>, or unsubscribe<https://gcc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACWGDTYXPXEPK2OAKBY6PI3RQNWRDANCNFSM4M3ZWWXA&data=02%7C01%7Crobert.podgorney%40inl.gov%7C85a17b2bfe9b49eb0e5a08d7f2f9bac2%7C4cf464b7869a42368da2a98566485554%7C0%7C0%7C637245026717829249&sdata=jevaqVRlrGDmTAywNZN0ug8MBdZrmkORs5MMOYCozTc%3D&reserved=0>.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/522035/) on e3d1e72 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15262/site)

This comment will be updated on new commits.

WilkAndy commented:
Thanks for your encouragement, guys :-)

a

WilkAndy committed:
comprehensive unit testing of Solver.  Fixes #15261

cpgr commented on modules/geochemistry/src/utils/EquilibriumGeochemicalSystem.C:
```diff
@@ -0,0 +1,1310 @@
+                   _constraint_value[i]);
+
+    // check that water is provided with correct meaning
+    if (name == "H2O")
```
Is this function called just at the start? If not, and it is called every iteration, we probably don't want to use a string comparison here as it is slow

WilkAndy replied:
Yea, it's just called in the constructor, so string comparison shouldn't be too expensive

cpgr commented on modules/geochemistry/src/utils/EquilibriumGeochemicalSystem.C:
```diff
@@ -0,0 +1,1310 @@
+    const std::string name = name_index.first;
+    const unsigned basis_ind = name_index.second;
+    const ConstraintMeaningEnum meaning = _constraint_meaning[basis_ind];
+    if (name == "H2O")
```
Same here. Do we have a 'H2O_index' to use instead?

WilkAndy replied:
Yea, this is just in the constructor, so probably OK.

cpgr commented on modules/geochemistry/src/utils/EquilibriumGeochemicalSystem.C:
```diff
@@ -0,0 +1,1310 @@
+            else if (_mgd.basis_species_charge[i] ==
+                     0.0) // certainly includes water, minerals and gases
+              continue;
+            else if (_in_algebraic_system[i]) // bulk moles must have been provided by user
```
else instead of else if?

WilkAndy replied:
I *think* this should be `else if`.   I'm happy to be wrong!  Did you work out the jacobian and think it should be `else` ?

cpgr replied:
I couldn't tell - you are probably correct!

cpgr commented on modules/geochemistry/src/utils/EquilibriumGeochemicalSystem.C:
```diff
@@ -0,0 +1,1310 @@
+            else if (_mgd.basis_species_charge[i] ==
+                     0.0) // certainly includes water, minerals and gases
+              continue;
+            else if (_constraint_meaning[i] == ConstraintMeaningEnum::ACTIVITY ||
```
Hard to tell but should this be else instead of else if? It might be just how it looks on GitHub

WilkAndy replied:
Again, i *think* this should be `else if`

cpgr reviewed (commented):
I only had some minor things. I think that it looks really good. We might find some things to improve on as you continue, but for now I think you have done a great job!

WilkAndy commented on modules/geochemistry/include/utils/EquilibriumGeochemicalSolver.h:
```diff
@@ -0,0 +1,144 @@
+                  std::stringstream & ss) const;
+
+  /**
+   * Progressively alger the initial-guess molalities for the algebraic system to attempt to reduce
```
Fixed.

WilkAndy committed:
Action added to solver and output equilibrium reaction solution.
The Output EquilibriumReactionSolverOutput instantiates a Solver, solves the specified system, and outputs results
The Action AddEquilibriumReactionSolverAction creates a small MOOSE model (since the mesh is irrelevant here) and creates a EquilibriumReactionSolverOutput object

Tests and examples added
Documentation modified

Fixes #15290

Note: because the temperature dependence of Debye-Huckel and equilibrium constants has not been included, the Gold expected output in these tests is very limited, and the documentation is currently incomplete.  For more discussion, Refs #15293

WilkAndy commented:
Hey @loganharbour .  This is getting ... monumental.   But I'm going to stop now as I feel this milestone is now complete.  Are you going to be brave and review it?

loganharbour commented:
I was actually waiting for a ping! Will get to it tomorrow.

On Thu, May 14, 2020 at 20:12 Andy Wilkins <notifications@github.com> wrote:

> Hey @loganharbour <https://github.com/loganharbour> . This is getting ...
> monumental. But I'm going to stop now as I feel this milestone is now
> complete. Are you going to be brave and review it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/15262#issuecomment-628984439>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIKINICFKYQXHMCE42IDSLRRSQHBANCNFSM4M3ZWWXA>
> .
>
-- 
Logan Harbour
Idaho National Laboratory Graduate Fellow
Department of Nuclear Engineering | Texas A&M University

WilkAndy commented:
Good luck, @loganharbour !

WilkAndy commented:
No comments, @loganharbour ?

moosebuild commented:
Job [Minimum gcc version](https://civet.inl.gov/job/526825/) on 18d5f7b : invalidated by @loganharbour

unrelated failure

WilkAndy commented:
Hey @loganharbour ..... hopefully you'll get to this monster soon.  I'm about to start another stage: #15330, woohoo

WilkAndy commented:
Hey @loganharbour .  I'm really hoping to get this merged soon.  I've just written another 2000 lines (for #15330) and it'll just keep building up and up to some evil monstrosity if we're not careful.

WilkAndy commented:
Bugging @loganharbour again !

permcody commented:
Hi @loganharbour - What's the latest on this? Is the Min GCC failure related to this PR or something else?

moosebuild commented:
Job [Minimum gcc version](https://civet.inl.gov/job/526825/) on 18d5f7b : invalidated by @loganharbour

Try this again

loganharbour committed:
Clear stringstream intead for min gcc

moosebuild commented:
Job [Test debug](https://civet.inl.gov/job/534441/) on 0719add : invalidated by @loganharbour

Unrelated?

loganharbour commented on modules/geochemistry/src/utils/EquilibriumGeochemicalSystem.C:
```diff
@@ -0,0 +1,1310 @@
+std::vector<Real>
+EquilibriumGeochemicalSystem::getAlgebraicVariableValues() const
+{
+  std::vector<Real> var;
```
Just a comment and not a change request: you know the size of these vectors ahead of time! You should size stuff like this at init wherever you can.

permcody replied:
Even better is to switch to std::array for fixed size "vectors". It can make a big difference as they are stack allocated, not heap allocated so you get better memory locality.

WilkAndy replied:
In this case, i don't the size of this vector, as it can change during a simulation.  There are other cases that are fixed at instantiation, however, and i can change those if you like.

WilkAndy replied:
Hmmm, i didn't understand about arrays.  I see their sizes have to be known at compile time, which is not true in this case, boohoo.  So i won't be changing any vectors to arrays

loganharbour commented on modules/geochemistry/doc/content/source/outputs/EquilibriumReactionSolverOutput.md:
```diff
@@ -0,0 +1,23 @@
+# EquilibriumReactionSolverOutput
+
+Usually this Output is added using the [EquilibriumReactionSolver action](EquilibriumReactionSolver/index.md).  The purpose of this Output is to solve an equilibrium reaction system and then output bulk compositions, molalities, activities, etc.  Some examples are:
```
Great addition 👍

loganharbour commented on modules/geochemistry/src/outputs/EquilibriumReactionSolverOutput.C:
```diff
@@ -0,0 +1,417 @@
+    _console << ss.str();
+
+  // retrieve information
+  const std::vector<Real> basis_molality = _egs.getSolventMassAndFreeMolalityAndMineralMoles();
+  std::vector<Real> basis_activity;
+  for (unsigned i = 0; i < _num_basis; ++i)
+    basis_activity.push_back(_egs.getBasisActivity(i));
+  std::vector<Real> basis_act_coef;
+  for (unsigned i = 0; i < _num_basis; ++i)
+    basis_act_coef.push_back(_egs.getBasisActivityCoefficient(i));
+  const std::vector<Real> bulk_moles = _egs.getBulkMoles();
+  std::vector<Real> eqm_molality;
+  for (unsigned j = 0; j < _num_eqm; ++j)
+    eqm_molality.push_back(_egs.getEquilibriumMolality(j));
+  std::vector<Real> eqm_act_coef;
+  for (unsigned j = 0; j < _num_eqm; ++j)
+    eqm_act_coef.push_back(_egs.getEquilibriumActivityCoefficient(j));
+  const std::vector<Real> eqm_SI = _egs.getSaturationIndices();
```
You should probably have getters for all of these values in the form of the entire vector, so that this simplifies to

```
const auto & basis_molality = _egs.getSolventMassAndFreeMolalityAndMineralMoles();
const auto & basis_activity = _egs.getBasisActivity();
const auto & basis_act_coef = _egs.getBasisActivityCoefficients();
...
```

WilkAndy replied:
OK, will do

WilkAndy replied:
Done

loganharbour commented on modules/geochemistry/src/utils/EquilibriumGeochemicalSolver.C:
```diff
@@ -0,0 +1,393 @@
+  const std::vector<Real> original_molality = _egs.getAlgebraicBasisValues();
+  ind = 0;
+  std::vector<unsigned> mol_order(_num_basis_in_algebraic_system);
+  std::iota(mol_order.begin(), mol_order.end(), ind++);
```
You do this a ton (7 times) on what seems to be the same type of vectors.

Make a helper for it?

WilkAndy replied:
Will look at it

WilkAndy replied:
done

loganharbour commented on modules/geochemistry/src/utils/EquilibriumGeochemicalSolver.C:
```diff
@@ -0,0 +1,393 @@
+    _is.setMaxStoichiometricIonicStrength(max_is0);
+    _abs_residual = computeResidual(_residual);
+    bool reducing_initial_molalities = (_abs_residual > _max_initial_residual);
+    while (reducing_initial_molalities)
```
Is this guaranteed to converge, always? could be stuck in an infinite loop

WilkAndy replied:
An infinite loop is impossible here, but a very long loop is possible.  i'll put in a counter to abort the loop if it goes on too long.

loganharbour reviewed (changes requested):
Minor comments. Some of the if statements in the solve like in the Jacobian are definitely hairy - but I have faith in them being correct 😬 . Per usual, unit tests are insane and comprehensive. I would definitely take a look [here](https://mooseframework.inl.gov/docs/coverage/geochemistry/) at the coverage if your unit test urges continue in the future.

I was holding off until this was passing - but I went ahead and took care of the gcc failure. I'll be around this evening if you get back in your morning and want it merged today!

WilkAndy commented on modules/geochemistry/doc/content/modules/geochemistry/tests_and_examples/morro.md:
```diff
@@ -85,4 +85,8 @@ MOOSE produces a total free charge of ????
 \end{equation}
 MOOSE produces the result of ????
```
Yea, i do have results, but don't want to put them into the documentation currently, because  we don't yet have accurate temperature dependence (waiting on a PR of @cpgr to be merged).  So we get reasonable agreement with other codes in most of the tests/examples, but not spectacular because of this.  I'll wait till we get accurate temp dependence....

loganharbour commented:
Last general comment: I know these species vectors are small in size, but you copy them a whole lot when trying to only read from them. In the future, you should try to do const references of them everywhere you can! In many cases, you already have const copies!

WilkAndy commented:
> Last general comment: I know these species vectors are small in size, but you copy them a whole lot when trying to only read from them. In the future, you should try to do const references of them everywhere you can! In many cases, you already have const copies!

OK, will do

WilkAndy commented:
> Minor comments. Some of the if statements in the solve like in the Jacobian are definitely hairy - but I have faith in them being correct 😬 . Per usual, unit tests are insane and comprehensive. I would definitely take a look [here](https://mooseframework.inl.gov/docs/coverage/geochemistry/) at the coverage if your unit test urges continue in the future.
> 
> I was holding off until this was passing - but I went ahead and took care of the gcc failure. I'll be around this evening if you get back in your morning and want it merged today!

Thanks for the link to the code coverage.  It's a shame that it doesn't include all those unit tests, boohoo.

loganharbour commented:
> > Minor comments. Some of the if statements in the solve like in the Jacobian are definitely hairy - but I have faith in them being correct 😬 . Per usual, unit tests are insane and comprehensive. I would definitely take a look [here](https://mooseframework.inl.gov/docs/coverage/geochemistry/) at the coverage if your unit test urges continue in the future.
> > I was holding off until this was passing - but I went ahead and took care of the gcc failure. I'll be around this evening if you get back in your morning and want it merged today!
> 
> Thanks for the link to the code coverage. It's a shame that it doesn't include all those unit tests, boohoo.

Darn. It really should. I'll look into that. We include framework unit tests in framework coverage... odd.

loganharbour commented on modules/geochemistry/src/utils/EquilibriumGeochemicalSystem.C:
```diff
@@ -0,0 +1,1310 @@
+  return var;
+}
+
+std::vector<Real>
+EquilibriumGeochemicalSystem::getAlgebraicBasisValues() const
+{
+  std::vector<Real> var;
+  for (unsigned a = 0; a < _num_basis_in_algebraic_system; ++a)
+    var.push_back(_basis_molality[_basis_index[a]]);
+  return var;
+}
```
This is the kind of change that I meant, by the way:
```suggestion
std::vector<Real>
EquilibriumGeochemicalSystem::getAlgebraicBasisValues() const
{
  std::vector<Real> var(_num_basis_in_algebraic_system);
  for (unsigned a = 0; a < _num_basis_in_algebraic_system; ++a)
    var[a] = _basis_molality[_basis_index[a]
  return var;
}
```

Don't worry about it right now, but I wanted to clear up the discussion!

WilkAndy replied:
Oh, i see, remove the push_back

WilkAndy committed:
all changes requested done.  Fixes #15290

WilkAndy commented:
Boohoo, has something changed in libmesh that's resulting in the framework error?

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/534922/) on 2ff6907 : invalidated by @loganharbour

Unrelated

WilkAndy commented:
Woohoo - all green - thanks for your stoicism, @loganharbour .
